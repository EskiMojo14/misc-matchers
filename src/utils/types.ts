import type {
  MatcherContext as JestMatcherContext,
  MatcherUtils as JestMatcherUtils,
  ExpectationResult,
  Tester as JestTester,
} from "expect";


// yuck :(
export type Tester = {
  bivariantHack(
    this: ThisParameterType<JestTester>,
    ...args: Partial<Parameters<JestTester>>
  ): ReturnType<JestTester>;
}["bivariantHack"];

export interface MatcherUtils {
  customTesters?: Array<Tester>;
  equals: (
    a: unknown,
    b: unknown,
    customTesters?: Array<Tester>,
    strictCheck?: boolean,
  ) => boolean;
  utils: Pick<
    JestMatcherUtils["utils"],
    | "matcherErrorMessage"
    | "matcherHint"
    | "RECEIVED_COLOR"
    | "printWithType"
    | "printReceived"
    | "DIM_COLOR"
    | "stringify"
    | "printExpected"
    | "printDiffOrStringify"
  >;
}

export type MatcherContext = Pick<
  JestMatcherContext,
  "isNot" | "promise" | "expand"
> &
  MatcherUtils;

export type MatcherFunction<Expected extends Array<unknown> = []> = (
  this: MatcherContext,
  value: unknown,
  ...expected: Expected
) => ExpectationResult;
