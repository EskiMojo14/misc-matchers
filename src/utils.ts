import type { MatcherHintOptions } from "jest-matcher-utils";
import type { MatcherUtils } from "expect";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isMock = (received: any): received is jest.Mock =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  received?._isMockFunction === true;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isSpy = (received: any): received is jasmine.Spy =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.all === "function" &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.count === "function";

export function ensureMockOrSpy(
  utils: MatcherUtils["utils"],
  received: unknown,
  matcherName: string,
  expectedArgument?: string,
  options?: MatcherHintOptions,
): asserts received is jest.Mock | jasmine.Spy {
  if (!isMock(received) && !isSpy(received)) {
    throw new Error(
      utils.matcherErrorMessage(
        utils.matcherHint(matcherName, undefined, expectedArgument, options),
        `${utils.RECEIVED_COLOR(
          "received",
        )} value must be a mock or spy function`,
        utils.printWithType("Received", received, utils.printReceived),
      ),
    );
  }
}

/**
 * Given a label, return a function which given a string,
 * right-aligns it preceding the colon in the label.
 */
export const getRightAlignedPrinter = (label: string) => {
  // Assume that the label contains a colon.
  const index = label.indexOf(":");
  const suffix = label.slice(index);

  return (string: string, isExpectedCall: boolean) =>
    (isExpectedCall
      ? `->${" ".repeat(Math.max(0, index - 2 - string.length))}`
      : " ".repeat(Math.max(index - string.length))) +
    string +
    suffix;
};
