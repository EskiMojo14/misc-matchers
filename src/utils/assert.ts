export function assert(
  condition: unknown,
  message: string | (() => string),
): asserts condition {
  if (!condition) {
    throw new Error(typeof message === "function" ? message() : message);
  }
}
