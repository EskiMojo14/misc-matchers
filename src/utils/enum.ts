export interface EnumLike {
  [key: string]: string | number;
  [index: number]: string;
}

export const getValidEnumValues = (obj: EnumLike) =>
  Object.values(obj).filter((v) => typeof obj[v] !== "number");
