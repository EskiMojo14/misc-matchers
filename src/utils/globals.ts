// to avoid having separate jest and vitest suites, we create common definitions
// then mock during setup with the appropriate globals
import type {
  describe as jestDescribe,
  it as jestIt,
  expect as jestExpect,
  jest,
} from "@jest/globals";
import type {
  describe as vitestDescribe,
  it as vitestIt,
  expect as vitestExpect,
  vi,
  Mock as VitestMock,
} from "vitest";
import type { Satisfies } from "./types.typetest";

interface Each {
  <T extends Array<any> | [any]>(
    cases: ReadonlyArray<T>,
  ): (name: string, fn: (...args: T) => void) => void;
  <T>(cases: ReadonlyArray<T>): (name: string, fn: (value: T) => void) => void;
}
export declare const describe: (typeof vitestDescribe | typeof jestDescribe) & {
  each: Each;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type TestEach = Satisfies<
  typeof jestDescribe.each | typeof vitestDescribe.each,
  Each
>;

export declare const expect: typeof vitestExpect | typeof jestExpect;
export declare const it: typeof vitestIt | typeof jestIt;

export type Mock<R = any, A extends Array<any> = Array<any>, T = any> =
  | jest.Mock<(this: T, ...args: A) => R>
  | VitestMock<A, R>;

export declare const fn: {
  (): Mock;
  <R, A extends Array<any>, T>(fn?: (this: T, ...args: A) => R): Mock<R, A, T>;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type TestFn = Satisfies<typeof jest.fn | typeof vi.fn, typeof fn>;

throw new Error("This file should be mocked");
