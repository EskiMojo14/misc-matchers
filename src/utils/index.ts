import type { MatcherHintOptions } from "jest-matcher-utils";
import type { MatcherUtils } from "expect";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isMock = (received: any): received is jest.Mock =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  received?._isMockFunction === true;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isSpy = (received: any): received is jasmine.Spy =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.all === "function" &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.count === "function";

export function ensureMockOrSpy(
  utils: MatcherUtils["utils"],
  received: unknown,
  matcherName: string,
  expectedArgument?: string,
  options?: MatcherHintOptions,
): asserts received is jest.Mock | jasmine.Spy {
  if (!isMock(received) && !isSpy(received)) {
    throw new Error(
      utils.matcherErrorMessage(
        utils.matcherHint(matcherName, undefined, expectedArgument, options),
        `${utils.RECEIVED_COLOR(
          "received",
        )} value must be a mock or spy function`,
        utils.printWithType("Received", received, utils.printReceived),
      ),
    );
  }
}
