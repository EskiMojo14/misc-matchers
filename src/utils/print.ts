/**
 * Given a label, return a function which given a string,
 * right-aligns it preceding the colon in the label.
 */
export const getRightAlignedPrinter = (label: string) => {
  // Assume that the label contains a colon.
  const index = label.indexOf(":");
  const suffix = label.slice(index);

  return (string: string, isExpectedCall: boolean) =>
    (isExpectedCall
      ? `->${" ".repeat(Math.max(0, index - 2 - string.length))}`
      : " ".repeat(Math.max(index - string.length))) +
    string +
    suffix;
};
