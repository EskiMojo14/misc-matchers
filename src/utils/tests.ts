/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import ansiRegex from "ansi-regex";
// ansi-styles is ESM - as a result we need to run our Jest tests with `--experimental-vm-modules` flag
import style from "ansi-styles";
import type { NewPlugin } from "pretty-format";

export const alignedAnsiStyleSerializer: NewPlugin = {
  serialize(val: string | Error): string {
    // Return the string itself, not escaped nor enclosed in double quote marks.
    return (val instanceof Error ? val.message : val).replace(
      ansiRegex(),
      (match) => {
        switch (match) {
          case style.inverse.open:
            return "<i>";
          case style.inverse.close:
            return "</i>";

          case style.bold.open:
            return "<b>";
          case style.dim.open:
            return "<d>";
          case style.green.open:
            return "<g>";
          case style.red.open:
            return "<r>";
          case style.yellow.open:
            return "<y>";
          case style.bgYellow.open:
            return "<Y>";

          case style.bold.close:
            return "</b>";
          case style.dim.close:
            return "</d>";
          case style.green.close:
            return "</g>";
          case style.red.close:
            return "</r>";
          case style.yellow.close:
            return "</y>";
          case style.bgYellow.close:
            return "</Y>";

          default:
            return match; // unexpected escape sequence
        }
      },
    );
  },
  test(val: unknown) {
    return typeof val === "string" || val instanceof Error;
  },
};
