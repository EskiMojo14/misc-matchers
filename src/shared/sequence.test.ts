/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { i } from "vitest/dist/reporters-QGe8gs4b.js";
import { alignedAnsiStyleSerializer } from "../utils/tests";

expect.addSnapshotSerializer(alignedAnsiStyleSerializer);

describe("toSatisfySequence", () => {
  it("works with an array", () => {
    expect([1, 2, 3]).toSatisfySequence(
      (x) => x === 1,
      (x) => x === 2,
      (x) => x === 3,
    );
    expect([1, 2, 3]).not.toSatisfySequence(
      (x) => x === 1,
      (x) => x === 3,
      (x) => x === 2,
    );

    expect([1, 2, 3]).toSatisfySequence(
      (x) => x === 1,
      (x) => x === 2,
    );

    expect(() => {
      expect([1, 2, 3]).toSatisfySequence(
        (x) => x === 1,
        (x) => x === 3,
        (x) => x === 2,
      );
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect([1, 2, 3]).not.toSatisfySequence(
        (x) => x === 1,
        (x) => x === 2,
      );
    }).toThrowErrorMatchingSnapshot();
  });
  it("works with other iterables", () => {
    expect(new Set([1, 2, 3])).toSatisfySequence(
      (x) => x === 1,
      (x) => x === 2,
      (x) => x === 3,
    );
    expect(new Set([1, 2, 3])).not.toSatisfySequence(
      (x) => x === 1,
      (x) => x === 3,
      (x) => x === 2,
    );

    expect(new Set([1, 2, 3])).toSatisfySequence(
      (x) => x === 1,
      (x) => x === 2,
    );

    expect(() => {
      expect(new Set([1, 2, 3])).toSatisfySequence(
        (x) => x === 1,
        (x) => x === 3,
        (x) => x === 2,
      );
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect(new Set([1, 2, 3])).not.toSatisfySequence(
        (x) => x === 1,
        (x) => x === 2,
      );
    }).toThrowErrorMatchingSnapshot();
  });
  it("fails with a non-iterable", () => {
    expect(() => {
      expect(0).toSatisfySequence((x) => x === 0);
    }).toThrowErrorMatchingSnapshot();
    expect(0).not.toSatisfySequence((x) => x === 0);
  });
  it("fails if not passed any predicates", () => {
    expect(() => {
      // @ts-expect-error testing invalid usage
      expect([]).toSatisfySequence();
    }).toThrowErrorMatchingSnapshot();
  });
  it("fails if the iterable is too short", () => {
    expect(() => {
      expect([0]).toSatisfySequence(
        (x) => x === 0,
        (x) => x === 1,
      );
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      expect(new Set([0])).toSatisfySequence(
        (x) => x === 0,
        (x) => x === 1,
      );
    }).toThrowErrorMatchingSnapshot();
  });
  it("throws if one of the predicates is not a function", () => {
    expect(() => {
      expect([0, 1]).toSatisfySequence(
        (x) => x === 0,
        // @ts-expect-error testing invalid usage
        1,
      );
    }).toThrowErrorMatchingSnapshot();
  });
});

describe("sequence", () => {
  it("works with an array", () => {
    expect({ array: [1, 2, 3] }).toEqual({
      array: expect.sequence(
        (x) => x === 1,
        (x) => x === 2,
        (x) => x === 3,
      ),
    });
    expect({ array: [1, 2, 3] }).toEqual({
      array: expect.not.sequence(
        (x) => x === 1,
        (x) => x === 3,
        (x) => x === 2,
      ),
    });

    expect({ array: [1, 2, 3] }).toEqual({
      array: expect.sequence(
        (x) => x === 1,
        (x) => x === 2,
      ),
    });

    expect(() => {
      expect({ array: [1, 2, 3] }).toEqual({
        array: expect.sequence(
          (x) => x === 1,
          (x) => x === 3,
          (x) => x === 2,
        ),
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ array: [1, 2, 3] }).toEqual({
        array: expect.not.sequence(
          (x) => x === 1,
          (x) => x === 2,
        ),
      });
    }).toThrowErrorMatchingSnapshot();
  });
  it("works with other iterables", () => {
    expect({ array: new Set([1, 2, 3]) }).toEqual({
      array: expect.sequence(
        (x) => x === 1,
        (x) => x === 2,
        (x) => x === 3,
      ),
    });
    expect({ array: new Set([1, 2, 3]) }).toEqual({
      array: expect.not.sequence(
        (x) => x === 1,
        (x) => x === 3,
        (x) => x === 2,
      ),
    });

    expect({ array: new Set([1, 2, 3]) }).toEqual({
      array: expect.sequence(
        (x) => x === 1,
        (x) => x === 2,
      ),
    });

    expect(() => {
      expect({ array: new Set([1, 2, 3]) }).toEqual({
        array: expect.sequence(
          (x) => x === 1,
          (x) => x === 3,
          (x) => x === 2,
        ),
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ array: new Set([1, 2, 3]) }).toEqual({
        array: expect.not.sequence(
          (x) => x === 1,
          (x) => x === 2,
        ),
      });
    }).toThrowErrorMatchingSnapshot();
  });
  it("fails with a non-iterable", () => {
    expect(() => {
      expect({ array: 0 }).toEqual({
        array: expect.sequence((x) => x === 0),
      });
    }).toThrowErrorMatchingSnapshot();
    expect({ array: 0 }).toEqual({
      array: expect.not.sequence((x) => x === 0),
    });
  });
  it("fails if not passed any predicates", () => {
    expect(() => {
      // @ts-expect-error testing invalid usage
      expect({ array: [] }).toEqual({ array: expect.sequence() });
    }).toThrowErrorMatchingSnapshot();
  });
  it("fails if the iterable is too short", () => {
    expect(() => {
      expect({ array: [0] }).toEqual({
        array: expect.sequence(
          (x) => x === 0,
          (x) => x === 1,
        ),
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ array: new Set([0]) }).toEqual({
        array: expect.sequence(
          (x) => x === 0,
          (x) => x === 1,
        ),
      });
    }).toThrowErrorMatchingSnapshot();
  });
  it("throws if one of the predicates is not a function", () => {
    expect(() => {
      expect({ array: [0, 1] }).toEqual({
        array: expect.sequence(
          (x) => x === 0,
          // @ts-expect-error testing invalid usage
          1,
        ),
      });
    }).toThrowErrorMatchingSnapshot();
  });
});

describe("toMatchSequence", () => {
  it("works with an array", () => {
    expect([1, 2, 3]).toMatchSequence(1, 2, 3);

    expect([1, 2, 3]).not.toMatchSequence(1, 3, 2);

    expect([1, 2, 3]).toMatchSequence(1, 2);

    expect(() => {
      expect([1, 2, 3]).toMatchSequence(1, 3, 2);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect([1, 2, 3]).not.toMatchSequence(1, 2);
    }).toThrowErrorMatchingSnapshot();
  });

  it("works with other iterables", () => {
    expect(new Set([1, 2, 3])).toMatchSequence(1, 2, 3);

    expect(new Set([1, 2, 3])).not.toMatchSequence(1, 3, 2);

    expect(new Set([1, 2, 3])).toMatchSequence(1, 2);

    expect(() => {
      expect(new Set([1, 2, 3])).toMatchSequence(1, 3, 2);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect(new Set([1, 2, 3])).not.toMatchSequence(1, 2);
    }).toThrowErrorMatchingSnapshot();
  });

  it("fails with a non-iterable", () => {
    expect(() => {
      expect(0).toMatchSequence(0);
    }).toThrowErrorMatchingSnapshot();
    expect(0).not.toMatchSequence(0);
  });
  it("fails if the iterable is too short", () => {
    expect(() => {
      expect([0]).toMatchSequence(0, 1);
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      expect(new Set([0])).toMatchSequence(0, 1);
    }).toThrowErrorMatchingSnapshot();
  });
});

describe("sequenceOf", () => {
  it("works with an array", () => {
    expect({ value: [1, 2, 3] }).toEqual({
      value: expect.sequenceOf(1, 2, 3),
    });
    expect({ value: [1, 2, 3] }).toEqual({
      value: expect.not.sequenceOf(1, 3, 2),
    });

    expect({ value: [1, 2, 3] }).toEqual({
      value: expect.sequenceOf(1, 2),
    });

    expect(() => {
      expect({ value: [1, 2, 3] }).toEqual({
        value: expect.sequenceOf(1, 3, 2),
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ value: [1, 2, 3] }).toEqual({
        value: expect.not.sequenceOf(1, 2),
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it("works with other iterables", () => {
    expect({ value: new Set([1, 2, 3]) }).toEqual({
      value: expect.sequenceOf(1, 2, 3),
    });
    expect({ value: new Set([1, 2, 3]) }).toEqual({
      value: expect.not.sequenceOf(1, 3, 2),
    });

    expect({ value: new Set([1, 2, 3]) }).toEqual({
      value: expect.sequenceOf(1, 2),
    });

    expect(() => {
      expect({ value: new Set([1, 2, 3]) }).toEqual({
        value: expect.sequenceOf(1, 3, 2),
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ value: new Set([1, 2, 3]) }).toEqual({
        value: expect.not.sequenceOf(1, 2),
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it("fails with a non-iterable", () => {
    expect(() => {
      expect({ value: 0 }).toEqual({
        value: expect.sequenceOf(0),
      });
    }).toThrowErrorMatchingSnapshot();
    expect({ value: 0 }).toEqual({
      value: expect.not.sequenceOf(0),
    });
  });
  it("fails if the iterable is too short", () => {
    expect(() => {
      expect({ value: [0] }).toEqual({
        value: expect.sequenceOf(0, 1),
      });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect({ value: new Set([0]) }).toEqual({
        value: expect.sequenceOf(0, 1),
      });
    }).toThrowErrorMatchingSnapshot();
  });
});
