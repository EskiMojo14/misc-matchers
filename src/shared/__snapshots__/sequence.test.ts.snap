// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sequence fails if not passed any predicates 1`] = `sequence requires at least one predicate`;

exports[`sequence fails if the iterable is too short 1`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 3</g>

<d>  Object {</b>
<g>-   "array": sequence<(x) => x === 0, (x) => x === 1>,</g>
<r>+   "array": Array [</g>
<r>+     0,</g>
<r>+   ],</g>
<d>  }</b>
`;

exports[`sequence fails if the iterable is too short 2`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 3</g>

<d>  Object {</b>
<g>-   "array": sequence<(x) => x === 0, (x) => x === 1>,</g>
<r>+   "array": Set {</g>
<r>+     0,</g>
<r>+   },</g>
<d>  }</b>
`;

exports[`sequence fails with a non-iterable 1`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 1</g>

<d>  Object {</b>
<g>-   "array": sequence<(x) => x === 0>,</g>
<r>+   "array": 0,</g>
<d>  }</b>
`;

exports[`sequence throws if one of the predicates is not a function 1`] = `predicate must be a function, but predicate at index 1 was "number"`;

exports[`sequence works with an array 1`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 5</g>

<d>  Object {</b>
<g>-   "array": sequence<(x) => x === 1, (x) => x === 3, (x) => x === 2>,</g>
<r>+   "array": Array [</g>
<r>+     1,</g>
<r>+     2,</g>
<r>+     3,</g>
<r>+   ],</g>
<d>  }</b>
`;

exports[`sequence works with an array 2`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 5</g>

<d>  Object {</b>
<g>-   "array": not.sequence<(x) => x === 1, (x) => x === 2>,</g>
<r>+   "array": Array [</g>
<r>+     1,</g>
<r>+     2,</g>
<r>+     3,</g>
<r>+   ],</g>
<d>  }</b>
`;

exports[`sequence works with other iterables 1`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 5</g>

<d>  Object {</b>
<g>-   "array": sequence<(x) => x === 1, (x) => x === 3, (x) => x === 2>,</g>
<r>+   "array": Set {</g>
<r>+     1,</g>
<r>+     2,</g>
<r>+     3,</g>
<r>+   },</g>
<d>  }</b>
`;

exports[`sequence works with other iterables 2`] = `
<d>expect(</b><r>received</g><d>).</b>toEqual<d>(</b><g>expected</g><d>) // deep equality</b>

<g>- Expected  - 1</g>
<r>+ Received  + 5</g>

<d>  Object {</b>
<g>-   "array": not.sequence<(x) => x === 1, (x) => x === 2>,</g>
<r>+   "array": Set {</g>
<r>+     1,</g>
<r>+     2,</g>
<r>+     3,</g>
<r>+   },</g>
<d>  }</b>
`;

exports[`toSatisfySequence fails if not passed any predicates 1`] = `toSatisfySequence requires at least one predicate`;

exports[`toSatisfySequence fails if the iterable is too short 1`] = `
<d>expect(</b><r>received</g><d>).</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>[0]</g> to have at least 2 items, but it only had 1.
`;

exports[`toSatisfySequence fails if the iterable is too short 2`] = `
<d>expect(</b><r>received</g><d>).</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>Set {0}</g> to have at least 2 items, but it only had 1.
`;

exports[`toSatisfySequence fails with a non-iterable 1`] = `
<d>expect(</b><r>received</g><d>).</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>0</g> to be an iterable
`;

exports[`toSatisfySequence throws if one of the predicates is not a function 1`] = `predicate must be a function, but predicate at index 1 was "number"`;

exports[`toSatisfySequence works with an array 1`] = `
<d>expect(</b><r>received</g><d>).</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>2</g> to satisfy predicate at index 1.

Full sequence so far: <r>1</g>, <r>2</g>
`;

exports[`toSatisfySequence works with an array 2`] = `
<d>expect(</b><r>received</g><d>).</b>not<d>.</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>[1, 2, 3]</g> not to satisfy sequence
`;

exports[`toSatisfySequence works with other iterables 1`] = `
<d>expect(</b><r>received</g><d>).</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>2</g> to satisfy predicate at index 1.

Full sequence so far: <r>1</g>, <r>2</g>
`;

exports[`toSatisfySequence works with other iterables 2`] = `
<d>expect(</b><r>received</g><d>).</b>not<d>.</b>toSatisfySequence<d>(</b><g>expected</g><d>)</b>

Expected <r>Set {1, 2, 3}</g> not to satisfy sequence
`;
